Add flow based node and edge disjoint paths.

jtorrents opened the pull request:
This is a flow based implementation of edge and node disjoint paths. We compute
the maximum flow between source and target on an auxiliary directed
network. This auxiliary network is the same used in the flow based edge and
node connectivity algorithms. See `networkx.algorithms.connectivity.utils` for details.
The saturated edges in the residual network after running the maximum flow algorithm
correspond to edge or node disjoint paths between source and target in the original network.

This function handles both directed and undirected graphs, and can use all flow
algorithms from NetworkX flow package.

Related to the discussion in #2053

jfinkels commented on networkx/algorithms/connectivity/tests/test_disjoint_paths.py:
```diff
@@ -0,0 +1,219 @@
+msg = "Assertion failed in function: {0}"
+
+
+def is_path(G, path):
```
There is an `is_path` in `networkx.utils`. (It may move to `networkx.is_simple_path`, see pull request #1921.)

jtorrents replied:
Nice. I like that implementation better. Now I also notice that forgot to check that edge disjoint paths are actually paths in the tests. Regarding #1921 note that edge disjoint paths are not necessarily simple (node disjoint paths are).

jtorrents commented:
The failure on AppVeyor is unrelated and quite strange.  
```
======================================================================
ERROR: test_kcutsets.test_configuration
----------------------------------------------------------------------
Traceback (most recent call last):
  File "c:\python34-x64\lib\site-packages\nose\case.py", line 198, in runTest
    self.test(*self.arg)
  File "c:\python34-x64\lib\site-packages\networkx\algorithms\connectivity\tests\test_kcutsets.py", line 150, in test_configuration
    deg_seq = nx.utils.create_degree_sequence(100,nx.utils.powerlaw_sequence)
  File "c:\python34-x64\lib\site-packages\networkx\utils\random_sequence.py", line 60, in create_degree_sequence
    "Exceeded max (%d) attempts at a valid sequence."%max_tries)
networkx.exception.NetworkXError: Exceeded max (50) attempts at a valid sequence.
```
It seems that `nx.utils.create_degree_sequence` could not come up with a valid power law degree sequence for 100 nodes in 50 attempts. Is there a way to restart the job in AppVeyor without making another commit? I logged in with github credentials but didn't see how to do it.

jtorrents commented:
Just added a header to the test file. Let's see if we see that error again.

jtorrents commented:
Two main changes: 1) add a cutoff parameter. The user can specify
the maximum number of paths to yield, but we also take advantage
of the fact the number of disjoint paths is bounded by the degree
of source and target to speed up the computation, especially in
sparse graphs. Only `edmonds_karp` and `shortest_augmenting_path`
support cutoff.

And 2) Use directly the residual network returned by flow algorithms
instead of the standard NetworkX API `nx.maximum_flow`: we skip a
function call and the extra work of building a full flow dict.

These changes, in sparse networks and for the `edmonds_karp` flow
function reduce the running time by 33%. Will post some benchmarks
shortly.

jtorrents commented:
Benchmark of the different flow algorithms for graphs with different topologies (random dense graphs, and sparse with right skewed degree distribution). As with connectivity algorithms, `edmons_karp` is the fastest by far on sparse networks and it's the default. Note that for this algorithm computing disjoint paths for 100 pairs of nodes only takes 10 times more than computing it for a single pair, this is because we reuse the auxiliary digraph and the residual network. The approximation is from `approximation.node_connectivity` and given that it's way faster than the exact flow based computation I also plan to add an API for getting the node disjoint paths from `approximation.node_connectivity`. 

```python
import time
from operator import itemgetter
from itertools import islice

import networkx as nx
from networkx.algorithms import flow
from networkx.algorithms import connectivity as conn
from networkx.algorithms.approximation import node_connectivity
from networkx.utils import powerlaw_sequence, create_degree_sequence

flow_funcs = dict(
    edmonds_karp=flow.edmonds_karp,
    preflow_push=flow.preflow_push,
    shortest_ap=flow.shortest_augmenting_path,
)

msg = "    {0}:\t{1:.3f} seconds to find {2} dijoint paths"
msg_p = "    {0}:\t{1:.3f} seconds for {2} pairs of nodes."

def generate_node_pairs(G):
    u = max(G, key=G.degree)
    for v, d in sorted(G.degree(), key=itemgetter(1), reverse=True):
        if u != v and v not in G[u]:
            yield u, v

def get_n_pairs_of_nodes(G, n=1):
    pairs = generate_node_pairs(G)
    if n == 1:
        return next(pairs)
    return list(islice(pairs, n))

def build_power_law(n):
    deg_seq = create_degree_sequence(n, powerlaw_sequence, 100)
    G = nx.Graph(nx.configuration_model(deg_seq))
    G.remove_edges_from(G.selfloop_edges())
    G.name = 'Power law configuration model: {0}'.format(n)
    return G

def benchmark_disjoint_paths():
    graphs = []
    print('Building test graphs. This may take a while.')
    for p in [0.2, 0.5]:
        G = nx.fast_gnp_random_graph(500, p)
        graphs.append(G)
    for n in [1000, 5000, 10000]:
        G = build_power_law(n)
        graphs.append(G)
    for G in graphs:
        print(nx.info(G))
        u, v = get_n_pairs_of_nodes(G, 1)
        pairs = get_n_pairs_of_nodes(G, 100)
        print("Computing edge disjoint_paths for a single pair of nodes")
        for fname, flow_func in sorted(flow_funcs.items()):
            start = time.time()
            p = list(nx.edge_disjoint_paths(G, u, v, flow_func=flow_func))
            total = time.time() - start
            print(msg.format(fname, total, len(p)))
        print("Computing edge disjoint_paths for {} pairs of nodes".format(len(pairs)))
        for fname, flow_func in sorted(flow_funcs.items()):
            start = time.time()
            H = conn.build_auxiliary_edge_connectivity(G)
            R = flow.build_residual_network(H, 'capacity')
            kwargs = dict(flow_func=flow_func, auxiliary=H, residual=R)
            for x, y in pairs:
                p = list(nx.edge_disjoint_paths(G, x, y, **kwargs))
            total = time.time() - start
            print(msg_p.format(fname, total, len(pairs)))
        print("Computing node disjoint paths for a single pair of nodes")
        for fname, flow_func in sorted(flow_funcs.items()):
            start = time.time()
            p = list(nx.node_disjoint_paths(G, u, v, flow_func=flow_func))
            total = time.time() - start
            print(msg.format(fname, total, len(p)))
        print("Computing node disjoint_paths for {} pairs of nodes".format(len(pairs)))
        for fname, flow_func in sorted(flow_funcs.items()):
            start = time.time()
            H = conn.build_auxiliary_node_connectivity(G)
            R = flow.build_residual_network(H, 'capacity')
            kwargs = dict(flow_func=flow_func, auxiliary=H, residual=R)
            for x, y in pairs:
                p = list(nx.node_disjoint_paths(G, x, y, **kwargs))
            total = time.time() - start
            print(msg_p.format(fname, total, len(pairs)))
        print("Computing node disjoint paths approximation for a single pair of nodes")
        start = time.time()
        p = node_connectivity(G, u, v)
        total = time.time() - start
        print(msg.format('approximation', total, p))
        print("Computing node disjoint paths approximation for {} pairs of nodes".format(len(pairs)))
        start = time.time()
        for x, y in pairs:
            p = node_connectivity(G, x, y)
        total = time.time() - start
        print(msg_p.format('approximation', total, len(pairs)))

if __name__ == '__main__':
    benchmark_disjoint_paths()
```
Output:
```
jtorrents@saturn:~/projects/benchmark$ python3 benchmark_disjoint_paths.py 
Building test graphs. This may take a while.
Name: fast_gnp_random_graph(500,0.2)
Type: Graph
Number of nodes: 500
Number of edges: 24881
Average degree:  99.5240
Computing edge disjoint_paths for a single pair of nodes
    edmonds_karp:	0.606 seconds to find 120 dijoint paths
    preflow_push:	0.550 seconds to find 120 dijoint paths
    shortest_ap:	0.529 seconds to find 120 dijoint paths
Computing edge disjoint_paths for 100 pairs of nodes
    edmonds_karp:	4.514 seconds for 100 pairs of nodes.
    preflow_push:	7.368 seconds for 100 pairs of nodes.
    shortest_ap:	6.439 seconds for 100 pairs of nodes.
Computing node disjoint paths for a single pair of nodes
    edmonds_karp:	0.978 seconds to find 120 dijoint paths
    preflow_push:	0.867 seconds to find 120 dijoint paths
    shortest_ap:	0.840 seconds to find 120 dijoint paths
Computing node disjoint_paths for 100 pairs of nodes
    edmonds_karp:	22.352 seconds for 100 pairs of nodes.
    preflow_push:	17.305 seconds for 100 pairs of nodes.
    shortest_ap:	12.839 seconds for 100 pairs of nodes.
Computing node disjoint paths approximation for a single pair of nodes
    approximation:	0.008 seconds to find 120 dijoint paths
Computing node disjoint paths approximation for 100 pairs of nodes
    approximation:	0.681 seconds for 100 pairs of nodes.
Name: fast_gnp_random_graph(500,0.5)
Type: Graph
Number of nodes: 500
Number of edges: 62257
Average degree: 249.0280
Computing edge disjoint_paths for a single pair of nodes
    edmonds_karp:	1.296 seconds to find 278 dijoint paths
    preflow_push:	1.379 seconds to find 278 dijoint paths
    shortest_ap:	1.373 seconds to find 278 dijoint paths
Computing edge disjoint_paths for 100 pairs of nodes
    edmonds_karp:	12.380 seconds for 100 pairs of nodes.
    preflow_push:	21.432 seconds for 100 pairs of nodes.
    shortest_ap:	18.902 seconds for 100 pairs of nodes.
Computing node disjoint paths for a single pair of nodes
    edmonds_karp:	3.293 seconds to find 278 dijoint paths
    preflow_push:	1.929 seconds to find 278 dijoint paths
    shortest_ap:	1.897 seconds to find 278 dijoint paths
Computing node disjoint_paths for 100 pairs of nodes
    edmonds_karp:	148.950 seconds for 100 pairs of nodes.
    preflow_push:	44.558 seconds for 100 pairs of nodes.
    shortest_ap:	37.357 seconds for 100 pairs of nodes.
Computing node disjoint paths approximation for a single pair of nodes
    approximation:	0.025 seconds to find 278 dijoint paths
Computing node disjoint paths approximation for 100 pairs of nodes
    approximation:	2.297 seconds for 100 pairs of nodes.
Name: Power law configuration model: 1000
Type: Graph
Number of nodes: 1000
Number of edges: 2859
Average degree:   5.7180
Computing edge disjoint_paths for a single pair of nodes
    edmonds_karp:	0.057 seconds to find 18 dijoint paths
    preflow_push:	0.214 seconds to find 18 dijoint paths
    shortest_ap:	0.068 seconds to find 18 dijoint paths
Computing edge disjoint_paths for 100 pairs of nodes
    edmonds_karp:	0.531 seconds for 100 pairs of nodes.
    preflow_push:	5.352 seconds for 100 pairs of nodes.
    shortest_ap:	1.312 seconds for 100 pairs of nodes.
Computing node disjoint paths for a single pair of nodes
    edmonds_karp:	0.094 seconds to find 16 dijoint paths
    preflow_push:	0.285 seconds to find 16 dijoint paths
    shortest_ap:	0.200 seconds to find 16 dijoint paths
Computing node disjoint_paths for 100 pairs of nodes
    edmonds_karp:	1.075 seconds for 100 pairs of nodes.
    preflow_push:	14.508 seconds for 100 pairs of nodes.
    shortest_ap:	5.537 seconds for 100 pairs of nodes.
Computing node disjoint paths approximation for a single pair of nodes
    approximation:	0.003 seconds to find 16 dijoint paths
Computing node disjoint paths approximation for 100 pairs of nodes
    approximation:	0.137 seconds for 100 pairs of nodes.
Name: Power law configuration model: 5000
Type: Graph
Number of nodes: 5000
Number of edges: 14722
Average degree:   5.8888
Computing edge disjoint_paths for a single pair of nodes
    edmonds_karp:	0.382 seconds to find 33 dijoint paths
    preflow_push:	0.914 seconds to find 33 dijoint paths
    shortest_ap:	0.671 seconds to find 33 dijoint paths
Computing edge disjoint_paths for 100 pairs of nodes
    edmonds_karp:	3.090 seconds for 100 pairs of nodes.
    preflow_push:	35.909 seconds for 100 pairs of nodes.
    shortest_ap:	11.664 seconds for 100 pairs of nodes.
Computing node disjoint paths for a single pair of nodes
    edmonds_karp:	0.736 seconds to find 33 dijoint paths
    preflow_push:	1.586 seconds to find 33 dijoint paths
    shortest_ap:	1.628 seconds to find 33 dijoint paths
Computing node disjoint_paths for 100 pairs of nodes
    edmonds_karp:	7.622 seconds for 100 pairs of nodes.
    preflow_push:	70.769 seconds for 100 pairs of nodes.
    shortest_ap:	35.052 seconds for 100 pairs of nodes.
Computing node disjoint paths approximation for a single pair of nodes
    approximation:	0.019 seconds to find 32 dijoint paths
Computing node disjoint paths approximation for 100 pairs of nodes
    approximation:	0.744 seconds for 100 pairs of nodes.
Name: Power law configuration model: 10000
Type: Graph
Number of nodes: 10000
Number of edges: 31983
Average degree:   6.3966
Computing edge disjoint_paths for a single pair of nodes
    edmonds_karp:	0.639 seconds to find 34 dijoint paths
    preflow_push:	1.365 seconds to find 34 dijoint paths
    shortest_ap:	0.835 seconds to find 34 dijoint paths
Computing edge disjoint_paths for 100 pairs of nodes
    edmonds_karp:	5.239 seconds for 100 pairs of nodes.
    preflow_push:	60.691 seconds for 100 pairs of nodes.
    shortest_ap:	18.316 seconds for 100 pairs of nodes.
Computing node disjoint paths for a single pair of nodes
    edmonds_karp:	1.129 seconds to find 33 dijoint paths
    preflow_push:	2.373 seconds to find 33 dijoint paths
    shortest_ap:	2.488 seconds to find 33 dijoint paths
Computing node disjoint_paths for 100 pairs of nodes
    edmonds_karp:	11.746 seconds for 100 pairs of nodes.
    preflow_push:	135.929 seconds for 100 pairs of nodes.
    shortest_ap:	82.232 seconds for 100 pairs of nodes.
Computing node disjoint paths approximation for a single pair of nodes
    approximation:	0.056 seconds to find 33 dijoint paths
Computing node disjoint paths approximation for 100 pairs of nodes
    approximation:	1.851 seconds for 100 pairs of nodes.
```

jtorrents commented:
I think this is ready for merging. Any comments/reviews welcome.

jtorrents commented:
I've updated the code, rebased on master, and squashed all commits. The failures from Appveyor are unrelated; it seems that we get a "permission denied" when trying to write to a temp file in `test_sparse6.py`.

dschult commented:
Yes - Appveyor changed something last week.  I'll open an issue for that.

Thanks for this!
